### ISSUE TITLE
Psalm infers float|int when using template for positive-int|0

### ISSUE THREAD

**Issue by asgrim** [vimeo/psalm#7235]
Given the following snippet https://psalm.dev/r/28b89bee3b

```php
<?php

/**
 * @psalm-pure
 * @psalm-template T as positive-int|0
 * @psalm-param T $depth
 */
function recursiveFn(int $depth = 1): void {
    if ($depth > 1) {
        recursiveFn($depth - 1);
    }
}
```

Psalm reports:

```
ERROR: ArgumentTypeCoercion - 10:21 - Argument 1 of recursiveFn expects 0|positive-int, parent type float|int provided
```

Given the exactly the same snippet, except not using the template https://psalm.dev/r/412e21cd88

```php
<?php

/**
 * @psalm-pure
 * @psalm-param positive-int|0 $depth
 */
function recursiveFn(int $depth = 1): void {
    if ($depth > 1) {
        recursiveFn($depth - 1);
    }
}
```

There are no (related) errors; presumably `$depth` is correctly identified as `positive-int|0` here.

**Comment by orklah** [comment]
indeed, the support for arithmetic operation on templates is very limited in Psalm. 

The case with two templates is handled here: https://github.com/vimeo/psalm/blob/97e6511fab6622c1ca434c665dd895e77a0eeca2/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php#L378
In a very limited way

But your case with a template and a known type is not even handled at all (and fallback into the default int|float result). Basically, we would need some recursive call to `ArithmeticOpAnalyzer::analyzeOperands` about here:
https://github.com/vimeo/psalm/blob/97e6511fab6622c1ca434c665dd895e77a0eeca2/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php#L447
Where each component of the Union contained in `$left_type_part->as` and `$right_type_part->as` would be checked against each other and then combined with TypeCombiner::combine to retrieve the result

**Comment by orklah** [comment]
Very related to #6747
